sway-core/src/type_system/engine.rs

```diff
@@ -205,6 +200,45 @@ impl TypeEngine {
         }
     }
 
+    /// Recursively resolve type arguments
+    #[allow(clippy::too_many_arguments)]
+    fn resolve_type_arguments(
```

Comment:
Would you restructure this so that we can more efficiently capture errors when we use this method? change the function signature to:

```rust
    fn resolve_type_arguments(
        &self,
        type_arguments: &mut [TypeArgument],
        decl_engine: &DeclEngine,
        enforce_type_arguments: EnforceTypeArguments,
        namespace: &mut Namespace,
        mod_path: &Path,
    ) -> CompileResult<()>
```

Instead of this being a recursive function, would you change `resolve_type_arguments` to iterate through the elements of `type_arguments`? Then at the end of `resolve_type_arguments`, will you return either an `ok((), warnings, errors)` or an `err(warnings, errors)`, depending on if `errors` is empty? This way we can append any warnings or errors found and return an `err` in that case, while still attempting to resolve each type argument.

Reply:
I suppose I could turn this into a stack based implementation, I was concerned that it would be harder to understand what it's doing at a glance but it would indeed make error handling clearer.

We have to evaluate the whole tree of arguments either way though.